### DISCUSSION TITLE
Why does Dictionary<TK,TV> enforce read-only iterators if they are indexed by key?

### DISCUSSION THREAD

**Discussion by BrianBohe** [dotnet/runtime#87706] [Q&A]
In previous versions of .net framework, the following code will throw an Exception:

```
public static void Main()
{
  var sb = new Dictionary<int, int>()
  {
    {1,1},
    {2,1},
  };
  foreach (var kv in sb)
  {
    kv.Value += 1;
    Console.WriteLine("Editing " + kv.Key + ": " + kv.Value);
  }
}
```
and now it won't even compile with the latest version of .net core. The error as can [see here](https://dotnetfiddle.net/XsDBRa) is: "Property or indexer 'System.Collections.Generic.KeyValuePair<int,int>.Value' cannot be assigned to -- it is read only"

What is the purpose behind avoiding iterators' value edition?

**Comment by nuiva** [comment]
To clarify, what would you expect the following to print?
```C#
public static void Main()
{
  var sb = new Dictionary<int, int>()
  {
    {1,1},
    {2,1},
  };
  foreach (var kv in sb)
  {
    kv.Value += 1;
    Console.WriteLine("Editing " + kv.Key + ": " + sb[kv.Key]);
  }
}
```
KeyValuePair<int,int> is just a simple struct containing two ints (which are copied by value), so any mutations to it wouldn't affect the underlying Dictionary even if it wasn't readonly.

**Reply by BrianBohe** [reply]
I was trying to give the smallest possible example and maybe this was too simple. What if we replace int with class T? A simple class that just has an integer as a field would work I guess

**Comment by huoyaoyuan** [comment]
> In previous versions of .net framework, the following code will throw an Exception:

On which version of .NET Framework and compiler it can compile? I don't think the rule on struct-returning property has changed. Maybe related to C# 5 foreach change?

**Reply by BrianBohe** [reply]
The example code compiles without problem using .net framework 4.7.2, does not compile with latest .net core version available

**Reply by svick** [reply]
@BrianBohe It does not compile for me when tageting .Net Framework 4.7.2:

```
> cat .\hwapp.csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net472</TargetFramework>
  </PropertyGroup>

</Project>
> cat .\Program.cs
using System;
using System.Collections.Generic;

class Program
{
    public static void Main()
    {
        var sb = new Dictionary<int, int>()
        {
            {1,1},
            {2,1},
        };
        foreach (var kv in sb)
        {
            kv.Value += 1;
            Console.WriteLine("Editing " + kv.Key + ": " + kv.Value);
        }
    }
}
> dotnet build
MSBuild version 17.5.0+6f08c67f3 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\code\tmp\hwapp\Program.cs(15,13): error CS0200: Property or indexer 'KeyValuePair<int, int>.Value' cannot be assigned to -- it is read only [C:\code\tmp\hwapp\hwapp.csproj]

Build FAILED.

C:\code\tmp\hwapp\Program.cs(15,13): error CS0200: Property or indexer 'KeyValuePair<int, int>.Value' cannot be assigned to -- it is read only [C:\code\tmp\hwapp\hwapp.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.22
```

And I don't see how it could compile, [`KeyValuePair<TKey,TValue>.Value` does not have a setter on .Net Framework 4.7.2](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.keyvaluepair-2.value?view=netframework-4.7.2).
